src/rfp/mod.rs

```diff
@@ -0,0 +1,278 @@
+}
+
+#[near(serializers=[borsh, json])]
+#[derive(Clone)]
+#[serde(tag = "rfp_body_version")]
+pub enum VersionedRFPBody {
+    V0(RFPBodyV0),
+}
+
+impl From<VersionedRFPBody> for RFPBodyV0 {
+    fn from(solution: VersionedRFPBody) -> Self {
+        match solution {
+            VersionedRFPBody::V0(v0) => v0,
+        }
+    }
+}
+
+impl From<RFPBodyV0> for VersionedRFPBody {
+    fn from(p: RFPBodyV0) -> Self {
+        VersionedRFPBody::V0(p)
+    }
+}
+
+impl VersionedRFPBody {
+    pub fn latest_version(self) -> RFPBodyV0 {
+        self.into()
+    }
+}
+
+pub fn get_subscribers(proposal_body: &RFPBodyV0) -> Vec<String> {
+    let result = [
+        get_text_mentions(proposal_body.description.as_str()),
+        get_text_mentions(proposal_body.summary.as_str()),
+    ]
+    .concat();
+    result
+}
+
+enum LinkedProposalChangeOperation {
+    Add,
+    Remove,
+}
+
+
+impl Contract {
+    fn assert_can_link_unlink_rfp(&self, rfp_id: Option<RFPId>) {
+        if let Some(rfp_id) = rfp_id {
+            let rfp: RFP = self
+                .rfps
+                .get(rfp_id.into())
+                .unwrap_or_else(|| panic!("RFP id {} not found", rfp_id))
+                .into();
+            require!(
+                rfp.snapshot.body.latest_version().timeline.is_accepting_submissions() || self.is_allowed_to_write_rfps(env::predecessor_account_id()),
+                format!("The RFP {} is not in the Accepting Submissions state, so you can't link or unlink to this RFP", rfp_id)
+            );
+        }
+    }
+
+    fn get_rfp_labels(&self, rfp_id: RFPId) -> HashSet<String> {
+        let rfp: RFP = self
+            .rfps
+            .get(rfp_id.into())
+            .unwrap_or_else(|| panic!("RFP id {} not found", rfp_id))
+            .into();
+        rfp.snapshot.labels
+    }
+
+    pub(crate) fn get_linked_proposals_in_rfp(&self, rfp_id: RFPId) -> HashSet<ProposalId> {
+        let rfp: RFP = self.get_rfp(rfp_id).into();
+        rfp.snapshot.linked_proposals
+    }
+
+    fn change_linked_proposal_in_rfp(&mut self, rfp_id: RFPId, proposal_id: ProposalId, operation: LinkedProposalChangeOperation) {
+        let mut rfp: RFP = self.get_rfp(rfp_id).into();
+        let mut linked_proposals = rfp.snapshot.linked_proposals.clone();
+        match operation {
+            LinkedProposalChangeOperation::Add => {
+                linked_proposals.insert(proposal_id);
+            }
+            LinkedProposalChangeOperation::Remove => {
+                linked_proposals.remove(&proposal_id);
+            }
+        }
+        let new_snapshot = RFPSnapshot {
+            editor_id: env::predecessor_account_id(),
+            timestamp: env::block_timestamp(),
+            labels: rfp.snapshot.labels,
+            body: rfp.snapshot.body,
+            linked_proposals: linked_proposals,
+        };
+        rfp.snapshot = new_snapshot;
+        rfp.snapshot_history.push(env::block_height() - 1);
```

Comment:
`block height - 1` is not guaranteed to exist. It is a valid situation when a block is skipped (not produced in time). Let's record the block height of the moment when we set the value (that also means that we will store the current snapshot block height in the list)